### PR TITLE
BUKKIT-2220 Player.getLanguage()

### PR DESCRIPTION
**Pull request by pschichtel**
This pull request adds the method getLanguage() to the Player interface to retrieve the Player's configured language.

Use cases:
- per-player translation
- language statistics

As discussed on IRC: This doesn't fix the lack of a proper localization system, but it allows the plugin developers to implement their solutions untill Bukkit provides one, without depending on reflection crap. And even if Bukkit finally adds a localization system, someone might still implement it in a different way as it might suit his project better. I planned to open-source my implementation as a small independent library.

Addresses [BUKKIT-2220](https://bukkit.atlassian.net/browse/BUKKIT-2220)

CraftBukkit PR: https://github.com/Bukkit/CraftBukkit/pull/889

### CODE REVIEW

File: src/main/java/org/bukkit/entity/Player.java [hunk 1]
```diff
@@ -554,4 +554,10 @@
      * @return The current allowed speed, from -1 to 1
      */
     public float getWalkSpeed();
+    
+    /**
+     * Gets the language this player has configured.
+     * @return The locale code of the language like en_US
```
Comment (jrtc27) [inline src/main/java/org/bukkit/entity/Player.java]: Maybe say ", such as" or "e.g." instead of the horrible "like"? :)

**Comment by jrtc27** [comment]
Seems like a good feature request to me.

**Comment by pschichtel** [comment]
updated the description

**Comment by ghost** [comment]
Was this implemented? I saw some Servers that have this..

**Comment by pschichtel** [comment]
Usually plugins do this by using Reflection

**Comment by riking** [comment]
Shall we add the the Javadocs the fact that the returned language is (or should be, based on the example) in [ISO 639](https://en.wikipedia.org/wiki/ISO_639) format?

Also, the CB PR link appears to be broken

**Comment by riking** [comment]
For that matter, where the heck is your forked Bukkit repo? It doesn't show up on your user page.

**Comment by pschichtel** [comment]
this won't get pulled, as Amaranth wanted to make a complete I18n API.

**Comment by pschichtel** [comment]
I've remove that fork a while back as I won't do (Craft)Bukkit PRs anymore

**Comment by riking** [comment]
Well, somehow the Bukkit commit lives on inside the Bukkit repo, but not the CraftBukkit commit, it seems

**Comment by pschichtel** [comment]
yep, interesting

**Comment by EdGruberman** [comment]
A complete language API is the goal, yes.  If you are not interested in pursuing such yourself, can you please confirm by closing this ticket?

**Comment by pschichtel** [comment]
I kept this open because there was no response given by an Bukkit dev yet.

And I can't close the CB PR has Github gives me an 500 error

**Comment by EdGruberman** [comment]
Yeah that 500 is odd business indeed.  We might have to ask github to do magic eventually for it.

Amaranth is a Bukkit Core Team developer so his vicarious word counts.  I've also talked at length with him and Wolvereness and overheard EvilSeph all talk to the goal of wanting a fully flushed language API anyways.

**Comment by pschichtel** [comment]
I did that myself with Amaranth and Wolverness (not 100% sure about him) in IRC, but here in this PR thread was no response, that's why it was still open.

**Comment by Sheigutn** [comment]
You can get the Minecraft language with:
Player p = YOUR_PLAYER;
Object ep = getMethod("getHandle", p.getClass()).invoke(p, (Object[]) null);
Field f = ep.getClass().getDeclaredField("locale");
f.setAccessible(true);
String language = (String) f.get(ep);

**Comment by natemort** [comment]
@Sheigutn You've always been able to get the language using reflection.  Reflection is no substitute for an API addition.

**Comment by pschichtel** [comment]
And don't forget to cache the method and field, because the lookup is really slow

and +1 to @evilmidget38

**Comment by Sheigutn** [comment]
Earlier you did not have to get it with reflection...
It was ((CraftPlayer) YOUR_PLAYER).getHandle().getLocale()... :3

**Comment by pschichtel** [comment]
@Sheigutn the d field containing the locale name in the locale object was private

**Comment by Sheigutn** [comment]
Oh, I remember you are right :3 But it was e ;)

**Comment by jrtc27** [comment]
@Sheigutn Casting to `CraftPlayer` is just as bad as having to resort to reflection. It relies on implementation details which cannot be relied upon, not to mention the fact that it takes a bit of time to set up due to the CraftBukkit jar having its packages relocated (and requires altering and compiling for every major release). The whole point is to have an implementation independent API, which is what this adds.

**Comment by Sheigutn** [comment]
@jrtc27 I know, I only wanted to mention it...

**Comment by jrtc27** [comment]
Ok, sorry, wasn't sure whether you were on the same page as everyone else or not.

**Comment by Sheigutn** [comment]
I didn't know if you knew this so I poste it.

Am 04.09.2013 um 00:12 schrieb James Clarke notifications@github.com:

**Comment by Sheigutn** [comment]
posted*

Am 04.09.2013 um 00:12 schrieb James Clarke notifications@github.com:

**Pull request closed without merging**